File: frontend/src/domain/closures.ts
// Roads the coordinator marked as cut (backend app/closures.py). Every route after one goes around it.

export interface RoadClosure {
  id: string
  lat: number
  lon: number
  radius_m: number
  note: string | null
  created_at: string
  /** Residents already leaving by a route through here when it was closed: call them again. */
  affected: string[]
}

/** Changes whenever the set of closures changes, so routes asked before are asked again. */
export const closuresKey = (closures: RoadClosure[]): string => closures.map((closure) => closure.id).join(',')

/** The closed stretch as a GeoJSON polygon: a circle of the closure's radius, on the ground. */
export function closureArea(closure: RoadClosure, steps = 32): { type: 'Polygon'; coordinates: number[][][] } {
  const metresPerDegreeLat = 110_570
  const metresPerDegreeLon = 111_320 * Math.cos((closure.lat * Math.PI) / 180)
  const ring = Array.from({ length: steps + 1 }, (_, i) => {
    const angle = (2 * Math.PI * i) / steps
    return [
      closure.lon + (closure.radius_m * Math.cos(angle)) / metresPerDegreeLon,
      closure.lat + (closure.radius_m * Math.sin(angle)) / metresPerDegreeLat,
    ]
  })
  return { type: 'Polygon', coordinates: [ring] }
}

File: backend/app/closures.py
"""Roads the coordinator marked as cut. Every route planned after a closure goes around it.

In memory, like the rest of the demo's state, and cleared by a demo reset. A closure is a point on
the road and a radius: routes avoid that circle (evacuation.avoid_polygon), and the route cache is
keyed by the closures in force (fingerprint), so a route from before a closure is never served.
"""

import hashlib
import uuid
from datetime import UTC, datetime

import shapely
from shapely.geometry.base import BaseGeometry

from . import geo
from .models import RoadClosure, RoadClosureRequest

_closures: dict[str, RoadClosure] = {}


def add(request: RoadClosureRequest, affected: list[str] | None = None) -> RoadClosure:
    closure = RoadClosure(
        **request.model_dump(), id=uuid.uuid4().hex[:8], created_at=datetime.now(UTC), affected=affected or []
    )
    _closures[closure.id] = closure
    return closure


def remove(closure_id: str) -> bool:
    return _closures.pop(closure_id, None) is not None


def all() -> list[RoadClosure]:
    return sorted(_closures.values(), key=lambda c: c.created_at)


def area_m() -> BaseGeometry | None:
    """Every closed stretch as one shape, in metres; None when no road is closed."""
    if not _closures:
        return None
    return shapely.union_all([geo.point_m(c.lon, c.lat).buffer(c.radius_m) for c in _closures.values()])


def fingerprint() -> str:
    """Changes whenever the closures in force change; empty when there are none."""
    if not _closures:
        return ""
    parts = sorted(f"{c.lon:.5f},{c.lat:.5f},{c.radius_m:g}" for c in _closures.values())
    return hashlib.sha1("|".join(parts).encode()).hexdigest()[:10]


def forget() -> None:
    _closures.clear()
